### ISSUE TITLE
SA-1 BW-RAM debugging

### ISSUE THREAD

**Issue by rgalland** [RedGuyyyy/usb2snesw#4]
Hi @RedGuyyyy , I am currently the SD2NES pro and have encountered a bug when running Kirby 3. I have narrowed it down to the BW-RAM (might even be from the iRAM since data gets copied form iRAM to BW-RAM). I have implemented the SNES myself with my FPGA board so I was able to simulate my board with the SD2SNES starting at the exact point where the memory corruption occurs. And guess what? The simulation does not show any corruption. I have been in touch with Ikari about it and he said I could use the USB2SNES to view the SA-1 RAM but going through the SD2SNES range from address 0xE00000. I have tried to export the contents and I think too much data is getting updated. Would it be possible to only visualize the iRAM (0x800 bytes) or the BW-RAM (0x8000 bytes with Birky 3)? I also tried to use the SA-1 debug stuff you added but to no avail.

**Comment by RedGuyyyy** [comment]
For cart RAM (BWRAM) In the MemoryViewer:
- Select "SD2SNES_RANGE" from the drop down box
- In the two boxes next to the [Export] button at the bottom type: E00000 and 8000
- Click [Export] if you want to write it to a file.  It's saved as export-SD2SNES_RANGE.bin (32KB if you used 8000 as the size) in the directory where the MemoryViewer is run from.


![cartRAM](https://user-images.githubusercontent.com/17366292/74597208-fac27180-500f-11ea-94f0-c21a6add9bcb.PNG)



Reading IRAM or really any non-PSRAM state would require FPGA support.  I think only the SA1 part of that is left in Ikari's files.  The other changes in the MCU, MSU, and top aren't there.  And debugging IRAM (DEBUG=1) was a bad hack for the non-PRO SD2SNES where 98-99% of the logic blocks were used with SA1.  To reduce logic I let it share a port with the SNES which created race conditions in the debug output when it was enabled.  Without changing the verilog there's no good way to read out IRAM without running code on the SA1 or on the SNES.

Nice work on implementing a SNES core.  Can you inject code to read out the state you need?  You could always read out IRAM from the SNES and write it to BWRAM and then use the MemoryViewer to read it from the SD2SNES.  You may be able to expand the BWRAM in the ROM header and write to a BWRAM location that isn't normally used.  I forgot how much space is supported by the SA1 implementation.  Unfortunately, only the base bit file on the non-PRO had space for the extra features like the dynamic hook and saving SNES RAM state so they aren't available in SA1.

**Comment by rgalland** [comment]
Thank you for your very comprehensive answer. I have applied the correct settings in the memory viewer and have been able to made further progress and managed to see where the data in the BW-RAM is written incorrectly. I still suspect the IRAM  is the problem but I have also noticed that the data is only wrong when reading from address 0xExxxxx (a bit of a coincidence from my previous message but unrelated) as there are previous reads from 0xCxxxxx and 0xDxxxxx. I have attached my notes with some SA-1 trace from bsnes:
[KirbysDreamLand3_SA1bug_notes.txt](https://github.com/RedGuyyyy/usb2snesw/files/4212538/KirbysDreamLand3_SA1bug_notes.txt)
If the problem is with address 0xExxxxx, can it be a problem with SA-1 bank selection? I checked sa1.v and it seems, though. The other problem could be the ROM mask. I have printed out the CMD registers but can't see where the ROM mask is. Would you be a able to point in the right direction?
![Kirby3_SD2SNES_CmdRegs](https://user-images.githubusercontent.com/27197663/74636841-1ab97880-5169-11ea-96ad-6536e6032942.png)

**Comment by RedGuyyyy** [comment]
I don't think this is a simple functional problem.  The game plays correctly on all my SNES including 2 original hardware and SuperNT with both original SD2SNES and SD2SNES PRO.  I used to have a PAL SNES which also worked fine on the non-PRO.  The PRO wasn't released at the time to test it.  Have you tried your SD2SNES on original hardware to see if it works?

The ROM mask is written to a register by the firmware.  I don't think it's available in the SD2SNES "SNESCMD" region at $2A00-2BFF: https://github.com/mrehkopf/sd2snes/blob/develop/verilog/sd2snes_base/mcu_cmd.v#L230

Are the traces for the cartridge bus long?  The SD2SNES and the original SNES hardware in general seems pretty sensitive to that.  I tried to make a breakout for a logic analyzer using a cheap cartridge adapter, but it only worked with original games and not the SD2SNES (any games).

Several of the SA1 games have races in their init code.  I tried using instruction breakpoints in the SA1 during development and certain code addresses would always result in stalls because the SNES would expect the SA1 to be in a certain state without using proper synchronization.  Is it possible your SNES is significantly slower or faster than an original SNES which could expose the race in the game code?

The firmware calculates a CRC from the ROM stored in PSRAM at game start.  This would randomly cause problems for one SA1 game, but Ikari made a change which seemed to help.  You could try delaying the start of execution or commenting out the call in the firmware to see if this is causing a problem: https://github.com/mrehkopf/sd2snes/blob/develop/src/snes.c#L266

It could be a bug in the SA1 verilog that somehow isn't exposed on any of my hardware, but not sure what suggestions to make on debugging it.  I did all the development on the non-PRO which was difficult due to the small FPGA.  The PRO will probably be easier to debug on if the state is exported.  If it fails quickly and consistently can you read anything out using the SNES?

**Comment by rgalland** [comment]
@RedGuyyyy, the problem is also showing on my PAL SNES. However, I have just found out that the problem is not the SD2SNES or my system: the culprit was the game all along. I got a new (J) and (U) version of the games and these two run fine! I can't believe I spent so much time on this. Although it is not entirely wasted. Just for history, I had been in touch with Ikari already because of this game and also because of Mario RPG that had glitchy tiles (see [https://github.com/mrehkopf/sd2snes/commit/71e7ec54098e38d4c7ea38617b2810675ac07fa3](url)). Ikary made a fix with Mario that worked but I was still left with the Kirby 3 problem. I then bought the Japanese cartridge and implemented the CIC which was the only bit left to do on my SNES and this allowed my to run the game.

Anyway, I can sleep at night now. If you are interested, I have posted some videos on my youtube channel to show the progress. In the not too distant future, I should release another one about SA-1 games since it is the last think I wanted to show. https://www.youtube.com/channel/UCA8jqaSHgyl7Iv7jkBRiW6Q?view_as=subscriber 

Thank you again for your time and Ikari's. I guess you can close this issue now.
